File: code/read_iworx.m
function [data, event] = read_iworx(folder)

% READ_IWORX reads and converts various IWORX datafiles into a
% FieldTrip-type data structure, which subsequently can be used for
% preprocessing or other analysis methods implemented in Fieldtrip
%
% Use as
%   [data, event] = read_iworx(folder)
% where folder contains a data (.txt) and marks (.txt) file
%
% data has the following nested fields:
%    .trial
%    .time
%    .label
%
%  event has the following nested fields:
%    .type
%    .sample
%    .value
%
% Copyright (C) 2022, Arjen Stolk


% check the input
txtfiles = dir([folder filesep '*.txt']);
for l = 1:numel(txtfiles)
  if ~isempty(strfind(txtfiles(l).name, 'Marks'))
    markfile = [folder filesep txtfiles(l).name];
  else
    datafile = [folder filesep txtfiles(l).name];
  end
end

% read the marks file
event = [];
try
  T = readtable(markfile, 'DurationType', 'text', 'VariableNamingRule', 'preserve');
  for e = 1:size(T,1)
    event(e).type = 'trig';
    event(e).sample = sum(sscanf(T.("Real Time"){e},'%f:%f:%f').*[3600;60;1]); % hh:mm:ss
    event(e).value  = T.("MarkValue"){e};
  end
catch
  fprintf('a problem arose extracting the events\n')
end

% read the data file
data = [];
try
  % header
  fid = fopen(datafile,'r');
  str = textscan(fid,'%s','Delimiter','\r');
  str = str{1};
  fclose(fid);
  data.label = split(str{1}, '	');
  idx = [];
  for l = 1:numel(data.label) % discard time channels
    if isempty(strfind(data.label{l}, 'Time'))
      idx = [idx l];
    end
  end
  data.label = data.label(idx);

  % data
  T = readtable(datafile, 'DurationType', 'text', 'VariableNamingRule', 'preserve');
  trl = 1;
  data.trial{1,trl} = [];
  data.time{1,trl}  = [];
  for smp = 1:size(T,1)
    if isequal(T.("TimeOfDay"){smp}, 'TimeOfDay')
      trl = trl+1;
      data.trial{1,trl} = [];
      data.time{1,trl}  = [];
    else
      data.trial{1,trl} = [data.trial{1,trl} T{smp,idx}'];
      data.time{1,trl}  = [data.time{1,trl} sum(sscanf(T.("TimeOfDay"){smp},'%f:%f:%f').*[3600;60;1])]; % hh:mm:ss
    end
  end
catch
  fprintf('a problem arose reading the data\n')
end

File: code/read_iworx.py
#!/usr/bin/env python

"""
--------------------------------------------------------
  READ_IWORX reads and converts various IWORX datafiles
  into a FieldTrip-type data structure

  Use as
    data, event = read_iworx(folder)
  where folder contains a data (.txt) and marks (.txt) file

  data has the following nested fields:
    .trial
    .time
    .label

  event has the following nested fields:
    .type
    .sample
    .value

  Copyright (C) 2022, Arjen Stolk
  --------------------------------------------------------
"""


import glob
import numpy as np


def read_iworx(folder):
    # check the input
    txtfiles = glob.glob(folder + '/*.txt')
    for l in txtfiles:
        if l.find('Marks') > 0:
            markfile = l
        else:
            datafile = l

    # read the marks file
    class Event(object):
        def __init__(self):
            self.type = []
            self.sample = []
            self.value = []
    event = Event()
    try:
        with open(markfile) as f:
            contents = f.readlines()
            for e in range(1, len(contents)):
                event.type.append('trig')
                h, m, s = contents[e].split("	")[contents[0].split(
                    '\t').index('Real Time')].split(':')
                event.sample.append(float(h) * 3600 + float(m) * 60 + float(s))
                event.value.append(contents[e].split(
                    "	")[contents[0].split('\t').index('MarkValue')])
    except:
        print('a problem arose extracting the events')

    # read the data file
    class Data(object):
        def __init__(self):
            self.trial = []
            self.time = []
            self.label = []
    data = Data()
    try:
        with open(datafile) as f:
            contents = f.readlines()
            # header
            data.label = contents[0].rstrip().split('\t')
            idx = []
            for i, l in enumerate(data.label):
                if l.find('Time'):
                    idx.append(i)
            data.label = [data.label[i] for i in idx]

            # data
            trl = 0
            data.trial.append([])
            data.time.append([])
            tmp = []
            for smp in range(1, len(contents)):
                if contents[smp].split('\t')[0] == 'TimeOfDay': # mark a new datablock/trial
                    data.trial[trl] = np.transpose(np.vstack(np.array(tmp)))
                    trl += 1
                    data.trial.append([])
                    data.time.append([])
                    tmp = []
                else: # read the data line by line
                    tmp.append([float(x) for i, x in enumerate(
                        contents[smp].rstrip().split('\t')) if i in idx])
                    if contents[0].split('\t')[0] == 'TimeOfDay': # if possible, add time information from TimeOfDay timestamps
                        h, m, s = contents[smp].split(
                            "	")[contents[0].split('\t').index('TimeOfDay')].split(':')
                        data.time[trl].append(
                            float(h) * 3600 + float(m) * 60 + float(s))
            data.trial[trl] = np.transpose(np.vstack(np.array(tmp)))
    except:
        print('a problem arose reading the data')

    return data, event
